core/node/state_keeper/src/keeper.rs

```diff
@@ -187,7 +189,10 @@ impl ZkSyncStateKeeper {
 
             // Finish current batch.
             if !updates_manager.l2_block.executed_transactions.is_empty() {
-                self.seal_l2_block(&updates_manager).await?;
+                if !self.should_create_l2_block {
+                    // l2 block has been already sealed
+                    self.seal_l2_block(&updates_manager).await?;
+                }
```

Comment:
This place is confusing with the proposed changes. The check above checks whether the latest block contains any transactions. AFAIU, if `should_create_l2_block` is true, then the check doesn't concern the *latest* block, but rather the previous one; the latest block isn't really started yet. So, a fictive block must be started in any case. IIUC, the current approach technically works because the previous block exists and is non-empty, but it looks hacky.

Reply:
see comment below.